### ISSUE TITLE
Universal way to get metadata about endpoint parameters.

### ISSUE THREAD

**Issue by voroninp** [dotnet/aspnetcore#42184]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

I am not sure whether this [issue](https://github.com/dotnet/aspnetcore/issues/41421) is about the same problem I have.

See also [my question](https://stackoverflow.com/questions/72618163/what-is-the-most-universal-way-to-get-metadata-about-defined-endpoint-parameters) on Stack Overflow.

I need a universal way to get metadata about endpoint parameters. It's not too difficult for controller actions but not that simple for the endpoints with delegate implementations.

### Describe the solution you'd like

Well, I expect that Endpoint's metadata contains `Parameters` property like `ControllerActionDescription` does.

### Additional context

See also [my question](https://stackoverflow.com/questions/72618163/what-is-the-most-universal-way-to-get-metadata-about-defined-endpoint-parameters) on Stack Overflow.

**Comment by davidfowl** [comment]
I don't see a why attached to that question, what are you trying to do?

**Comment by voroninp** [comment]
No, I use `IHttpContextAccessor` to get `HttpContext` and then I resolve the id:

```
if (!TryGetOrderId(context, attribute.OrderIdParameterName, out Guid orderId))
{
    return;
}
```

The methods are:
```
private bool TryGetOrderId(AuthorizationHandlerContext context, string paramName, out Guid orderId) =>
    TryGetOrderIdAsResource(context, out orderId)
    || TryGetOrderIdFromRouteData(paramName, out orderId) 
    || TryGetOrderIdFromQuery(paramName, out orderId);

private bool TryGetOrderIdFromQuery(string paramName, out Guid orderId)
{
    var httpContext = _httpContextAccessor.HttpContext;
    if (httpContext is null)
    {
        throw new InvalidOperationException("No active HttpConext");
    }
    var query = httpContext.Request.Query;
    var valuesFromQuery = query[paramName];
    if (valuesFromQuery.Count != 1)
    {
        orderId = Guid.Empty;
        return false;
    }

    return Guid.TryParse(valuesFromQuery.FirstOrDefault(), out orderId);
}
```

My concern here is that resource and route parameters are defined by the developer, yet the query parameters can be added by the requestor, and thus I would like to ensure this parameter is actually something endpoint is aware of.

Let's imagine there's an endpoint which accepts `orderId` as part of  body:

```
public sealed record Bar(Guid OrderId);

[OwnerOfOrder]
public IActionResult Foo(Bar payload) { ... }

[OwnerOfOrder]
public IActionResult Foo2([FromQuery(Name = "orderId")]Guid id) { ... }
```

`OwnerOfOrder` inherits from `Authorize` and just specifies the policy which has a requirement processed by the handler.

For `Foo` endpoint handler won't resolve `orderId` from the route, so authorization will always fail unless request contains additional query parameter `orderId`. But code does not use it, it works with the id in `payload`, so the values can differ.

Yet for `Foo2` it's safe to assume that `orderId` parameter will be bound to `id` and it will be used in code below.

This is a contrived example, but I hope it explains the `WHY`.

**Comment by captainsafia** [comment]
@voroninp We don't store any information about the parameter in the endpoint metadata at the moment. Your best bet is to leverage the strategy you mentioned in the StackOverflow issue and check the `ParamterInfo` of the method similar to what we do in the API description provider.

BTW, it might not help your particular scenario, but we introduced support for an `IEndpointMetadataProvider` in the .NET 7 that can be used to allow parameter/return types to annotate metadata with information. 

```csharp
class IdFromQueryOrRoute : IEndpointParameterMetadataProvider
{
  public static void PopulateMetadata(EndpointParameterMetadataContext parameterContext)
  {
    if (parameterContext.Parameter.HasFromQueryAttribute())
    {
      parameterContext.EndpointMetadata.Add(new FooMetadata())
    }
  }
}
```

**Comment by captainsafia** [comment]
I think `IEndpointParameterMetadataProvider` is a sufficient API here. Instead of endpoint's storing `ParmaterInfo` for each as metadata that is then derived in filters/handlers, the parameters themselves should project whatever information is necessary into metadata via the `IEndpointParameterMetadataProvider.

**Comment by davidfowl** [comment]
I think the bigger question is if we choose to invent a metadata model that encapsulates the decisions that happen inside of the RequestDelegateFactory. Right now we hand out the MethodInfo and you have to duplicate our logic (this is what the ApiExplorer implementation does as well).

**Comment by mitchdenny** [comment]
Removing the Needs Attention tag from this issue since I think that its on the backlog appropriately for now if we choose to emit more metadata which represents the binding decisions that were made in either RDF or RDG.
